### ISSUE TITLE
Performance issue

### ISSUE THREAD

**Issue by z0al** [lddubeau/saxes#14]
## Background

I used saxes to build [feedify][feedify] (RSS parsing package). But recently when I used feedify in one of my projects I noticed a bad parsing performance. 

## The problem

The browser window would freeze for a second for two when parsing, for example, feeds from [overreacted.io][overreacted].

I tried to investigate what is wrong with my implementation but honestly, I was just trying to guess things (I could not do proper profiling). After I failed I thought saxes might have something to do with this especially after trying other RSS parsing packages like [rss-parser][rss-parser] (it depends on `xml2js` which depends on `sax`). 

I tried both feedify and rss-parser with the same [input][overreacted] and there were no issues when using rss-parser. So I decided to give `sax` a shot.

Shockingly, replacing `saxes` with `sax` actually fixed my issue.

That being said, I've no idea why that was the case. 

## Things that might help

Here is the [source code of my parser][diff] before and after using `sax`.  

I was using saxes v3.1.7 (I [don't see any actual code change][compare] in v3.1.9)

> Note: the result isn't affected by changing the handlers (e.g. `onopentag`) to promises.

## Screenshots

I recorded the performance of running with both `saxes` and `sax` versions and here are screenshots of both:

### With saxes

![saxes](https://user-images.githubusercontent.com/12673605/59147065-81de3500-89ff-11e9-8820-a099fc415250.png)

### With sax

![sax](https://user-images.githubusercontent.com/12673605/59147066-860a5280-89ff-11e9-9981-2abee8ae4f8a.png)

## Environment 

Browser: Brave (Chromium: 73.0.3683.86) 

## Runner

Here is the script I used to run the parser.

```javascript
import http from 'ky';
import { parse } from 'feedify';

async function run(url){
	const resp = await http.get(url);
	
	if (resp.ok) {
		const xml = await resp.text();
	
		await parse(xml);
	}
}

run('https://overreacted.io/rss.xml')
```

[feedify]: https://github.com/z0al/feedify
[overreacted]: https://overreacted.io/rss.xml 
[compare]: https://github.com/lddubeau/saxes/compare/v3.1.7...master
[rss-parser]: https://github.com/bobby-brennan/rss-parser
[diff]: https://github.com/z0al/feedify/compare/v0.3.2...v0.3.3?diff=split#diff-ef3b2f5f79e5770376c2c2259e2d853f

**Comment by lddubeau** [comment]
Thank you for the report. I'm very much pressed for time right now but I did a very quick check with the RSS from `overreacted.io`, using the `null-parser.js` example bundled with saxes , and I used an equivalent null parser with sax. Sax was definitely faster on the same input. I noticed the XML from `overreacted.io` contains *a truckload* of XML entities. On a hunch I replaced them all with plain text. Once I did that, saxes was much faster than sax.

Right now my going hypothesis is that saxes is significantly slower than sax at parsing XML entities. My claims regarding the relative performance of sax and saxes that I have in the README are backed with actual tests but those tests only had a few XML entities so the issue with parsing entities did not show up when I ran those performance tests.

I can see ways to reorganize the validation logic which may help improve how fast XML entities get parsed. Right now, the logic is such that saxes verifies as it is reading the entity name that each character of an XML entity name is allowed. It may be doable to flip the logic so that saxes reads the name, tries to resolve it, and if it cannot resolve it, *then* checks whether the characters of the name are okay so as to produce an informative error message. The cost of checking the characters of an entity name would be spent only when the name does not resolve.

**Comment by z0al** [comment]
No worries, take your time.

**Comment by lddubeau** [comment]
@z0al My initial impression was wrong. The issue is not the checks performed on the entity names. The issue was with the check for `]]>` in character data, which XML disallows. (See https://www.w3.org/TR/xml/#syntax )

I've rewritten the logic that checks for `]]>` and pushed the changes to the branch https://github.com/lddubeau/saxes/tree/feature/improve-parsing-speed When I ran tests here, it was faster than sax. Can you try it?

**Comment by z0al** [comment]
Thanks for the update. I will try it in the morning

**Comment by z0al** [comment]
Yep, that was definitely faster. Thanks for the awesome work.

**Comment by lddubeau** [comment]
The fix is published in 3.1.10, which I've just released now.
